nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/main/java/org/apache/nifi/py4j/PythonProcess.java

```diff
@@ -0,0 +1,293 @@
+//      - Appears to still be issues with timeout in nifi when set to 10 secs... initially will probably just leave default of 0sec.
+
+public class PythonProcess {
+    private static final Logger logger = LoggerFactory.getLogger(PythonProcess.class);
+    private static final String PYTHON_CONTROLLER_FILENAME = "Controller.py";
+
+    private final PythonProcessConfig processConfig;
+    private final ControllerServiceTypeLookup controllerServiceTypeLookup;
+    private final File virtualEnvHome;
+    private GatewayServer server;
+    private PythonController controller;
+    private Process process;
+    private NiFiPythonGateway gateway;
+    private final Map<String, Boolean> processorPrefersIsolation = new ConcurrentHashMap<>();
+
+
+    public PythonProcess(final PythonProcessConfig processConfig, final ControllerServiceTypeLookup controllerServiceTypeLookup, final File virtualEnvHome) {
+        this.processConfig = processConfig;
+        this.controllerServiceTypeLookup = controllerServiceTypeLookup;
+        this.virtualEnvHome = virtualEnvHome;
+    }
+
+    public PythonController getController() {
+        return controller;
+    }
+
+    public void start() throws IOException {
+        // TODO: Look into using configured TLS Certs to make this secure by default.
+        final ServerSocketFactory serverSocketFactory = ServerSocketFactory.getDefault();
+        final SocketFactory socketFactory = SocketFactory.getDefault();
+
+        final int timeoutMillis = (int) processConfig.getCommsTimeout().toMillis();
+        final String authToken = null;
+        final CallbackClient callbackClient = new CallbackClient(GatewayServer.DEFAULT_PYTHON_PORT, GatewayServer.defaultAddress(), authToken,
+            50000L, TimeUnit.MILLISECONDS, socketFactory, false, timeoutMillis);
+
+        final JavaObjectBindings bindings = new JavaObjectBindings();
+        gateway = new NiFiPythonGateway(bindings, null, callbackClient);
+        gateway.startup();
+
+        server = new NiFiGatewayServer(gateway,
+            0,
+            GatewayServer.defaultAddress(),
+            timeoutMillis,
+            timeoutMillis,
+            Collections.emptyList(),
+            serverSocketFactory);
+        server.start();
+
+        final int listeningPort = server.getListeningPort();
+
+        setupEnvironment();
+        this.process = launchPythonProcess(listeningPort);
+
+        final StandardPythonClient pythonClient = new StandardPythonClient(gateway);
+        controller = pythonClient.getController();
+
+        final long timeout = System.currentTimeMillis() + TimeUnit.SECONDS.toMillis(60L);
+        Exception lastException = null;
+        boolean pingSuccessful = false;
+        while (System.currentTimeMillis() < timeout) {
+            try {
+                final String pingResponse = controller.ping();
+                pingSuccessful = "pong".equals(pingResponse);
```

Comment:
Yeah, I believe this can be clarified. I like the String over boolean because `boolean` can also be inferred from just about anything - null or not, 0 vs 1, etc. and may not convey the intent as clearly. But I do think that we should document in the interface very clearly that it is expected to return the String value `pong`.